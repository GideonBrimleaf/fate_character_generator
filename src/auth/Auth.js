import React, { useEffect, useState } from 'react';
import { auth } from '../lib/fire';

export const AuthContext = React.createContext();

export const AuthProvider = ({ children }) => {
  const [currentUser, setCurrentUser] = useState(null);
  const [pending, setPending] = useState(true);

  useEffect(() => {
    // TODO - this observer is firing and triggering state
    // changes in test renders which can't be handled with "act",
    // this is creating console warnings in test runs
    auth.onAuthStateChanged((user) => {
      setCurrentUser(user)
      setPending(false)
    });
  }, []);

  if (pending) {
    return <>Loading...</>
  }

  return (
    <AuthContext.Provider
      value={{
        currentUser,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};
