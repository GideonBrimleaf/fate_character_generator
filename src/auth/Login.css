.login-input:-webkit-autofill,
.login-input:-webkit-autofill:hover, 
.login-input:-webkit-autofill:focus,
textarea:-webkit-autofill,
textarea:-webkit-autofill:hover,
textarea:-webkit-autofill:focus,
select:-webkit-autofill,
select:-webkit-autofill:hover,
select:-webkit-autofill:focus {
  -webkit-text-fill-color: white;
  -webkit-box-shadow: 0 0 0px 1000px #282c34 inset;
  font-size: x-large;
}

.login-input::placeholder {
  text-align: left;
  font-size: x-large;
  color: white;
}

.login-input {
  min-height: 2.5vh;
  background-color: #282c34;
  border-top-style: none;
  border-right-style: none;
  border-left-style: none;
  border-bottom-color: white;
  caret-color: white;
  color: white;
  font-size: x-large;
}

.login-input[type="email"]:focus {
  outline: none;
}

.login-input[type="email"]:focus:hover {
  outline: none;
}

.login-input[type="password"]:focus {
  outline: none;
}

.login-input[type="password"]:focus:hover {
  outline: none;
}