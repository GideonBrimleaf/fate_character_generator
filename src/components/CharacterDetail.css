caption {
  font-weight: bold;
}

.character-sheet{
  color: black;
  background-color: lightcyan;
  width: 80vw;
  max-height: 90vh;
  margin-bottom: 2%;
}

.character-stats {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  margin-bottom: 20px;
}

.character-sheet-header {
  margin: 0px;
}

.character-sheet-text {
  font-size: calc(12px + 0.75vw);
}

.character-sheet-item-primary {
  margin-bottom: 0px;
  flex: 1;
  padding: 0px 0px 0px 10%;
}

.character-sheet-item-secondary {
  padding: 0px 10% 0px 0px;
}

.character-sheet-item-refresh {
  margin-right: 11%;
  margin-top: 5%;
}

.character-name {
  font-size: calc(30px + 1vw);
  padding-right: 20%;
  padding-top: 1%;
}

.stunt-header {
  text-align: left;
}

.refresh-header {
  text-align: center;
  display: inline;
}