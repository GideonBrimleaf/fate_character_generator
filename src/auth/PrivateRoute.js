import React, { useContext } from "react";
import { Route, Redirect } from "react-router-dom";
import { AuthContext } from "./Auth";

const PrivateRoute = ({ component: RouteComponent, data, ...rest }) => {
  const {currentUser} = useContext(AuthContext);
  return (
    <Route
      {...rest}
      render={routeProps =>
        !!currentUser ? (
          <RouteComponent {...routeProps} {...data} />
        ) : (
          <Redirect to={"/login"} />
        )
      }
    />
  );
};


export default PrivateRoute