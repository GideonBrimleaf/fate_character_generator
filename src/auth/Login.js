import React, { useCallback, useContext } from "react";
import { withRouter, Redirect } from "react-router";
import { auth } from "../fire";
import { AuthContext } from "./Auth.js";
import './Login.css'

const Login = ({ history }) => {
  const handleLogin = useCallback(
    async event => {
      event.preventDefault();
      const { email, password } = event.target.elements;
      try {
        await auth.signInWithEmailAndPassword(email.value, password.value);
        history.push("/");
      } catch (error) {
        alert(error);
      }
    },
    [history]
  );

  const { currentUser } = useContext(AuthContext);

  if (currentUser) {
    return <Redirect to="/" />;
  }

  return (
    <div>
      <h1>Log In</h1>
      <form onSubmit={handleLogin}>
        <input className="login-input" name="email" type="email" placeholder="Email:" />
        <input className="login-input" name="password" type="password" placeholder="Password:" />
        <button className="primary-button" type="submit">Log in</button>
      </form>
    </div>
  );
};

export default withRouter(Login);