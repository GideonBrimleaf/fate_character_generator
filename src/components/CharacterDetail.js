import React from 'react'
import './CharacterDetail.css'
import './CharacterStatBlock.css'
import CharacterStatBlock from './CharacterStatBlock'

const CharacterDetail = (props) => {

  const foundCharacter = props.characters.find(character => {
    return parseInt(props.match.params.characterId) === character.id
  })
  
  const characterStunts = foundCharacter.stunts.map(stunt => {
    return (
      <li key={stunt} className="align-left">{ stunt }</li>
      )
  })

  return (
    <main className="character-sheet">
      <section className="character-stats">
        <h1 className="character-name character-sheet-item-primary">{ foundCharacter.name }</h1>
        <div className="character-sheet-item-refresh">
          <h5 className="character-sheet-header refresh-header">Refresh</h5>
          <p className="refresh-stat">{ foundCharacter.refresh }</p>
        </div>
      </section>
      <section className="character-stats">
        <article className="character-sheet-item-primary">
          <CharacterStatBlock characterStatCategory="Aspects" characterStats={ foundCharacter.aspects } />
        </article>
        <article className="character-sheet-item-secondary">
          <CharacterStatBlock characterStatCategory="Approaches" characterStats={ foundCharacter.approaches } />
        </article>
      </section>
      <section className="character-stats">
        <article className="character-sheet-item-primary">
          <h5 className="character-sheet-header stunt-header">Stunts</h5>
          <ul className="stat-block">
            { characterStunts }
          </ul>
        </article>
      </section> 
      <section className="character-stats">
        <article className="character-sheet-item-primary">
          <CharacterStatBlock characterStatCategory="Consequences" characterStats={ foundCharacter.consequences } />
        </article>
      </section>
    </main>
  )
}
export default CharacterDetail
