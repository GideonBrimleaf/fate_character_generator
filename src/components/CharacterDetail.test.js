import React from 'react';
import { render, act } from '@testing-library/react';

import { AuthContext } from '../auth/Auth';
import CharacterDetail from './CharacterDetail';

it('renders a loading message is character is not found', () => {
  // Arrange
  const props = {
    match: {
      params: {
        characterId: '1234',
      }
    },
    characters: [],
  }

  // Act
  const result = render(
    <AuthContext.Provider value={{ currentUser: null }}>
      <CharacterDetail {...props} />
    </AuthContext.Provider>
  );

  // Assert
  expect(result.asFragment()).toMatchSnapshot();
});

it('renders character when a character is found', async () => {
  // Arrange
  const props = {
    match: {
      params: {
        characterId: '1234',
      }
    },
    characters: [
      {
        id: 1234,
        name: 'expected name',
        stunts: [],
        aspects: {expectedAspectName: 'expected aspect value'},
        approaches: {expectedApproachName: 'expected appraoch value'},
        consequences: {expectedConName: 'expected con value'},
      }
    ],
  }

  let result;

  // Act
  result = render(
    <AuthContext.Provider value={{ currentUser: null }}>
      <CharacterDetail {...props} />
    </AuthContext.Provider>
  );

  expect(result).toBeDefined();

  // Assert
  expect(result.asFragment()).toMatchSnapshot();
});