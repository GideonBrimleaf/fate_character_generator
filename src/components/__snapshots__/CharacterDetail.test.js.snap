// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a loading message is character is not found 1`] = `
<DocumentFragment>
  <p>
    Loading
  </p>
</DocumentFragment>
`;

exports[`renders character when a character is found 1`] = `
<DocumentFragment>
  <main
    class="character-sheet character-sheet-text top-padding"
  >
    <section
      class="character-stats"
    >
      <h1
        class="character-name character-sheet-item-primary"
      >
        expected name
      </h1>
      <h5
        class="character-sheet-header refresh-header character-sheet-text"
      >
        Refresh:
      </h5>
    </section>
    <section
      class="character-stats"
    >
      <article
        class="character-sheet-item-primary"
      >
        <table
          class="padded-table"
        >
          <caption
            class="table-header"
          >
            Aspects
          </caption>
          <tbody>
            <tr>
              <td
                class="stat-name"
              >
                High Concept:
              </td>
              <td
                class="primary-stat-description"
              />
            </tr>
            <tr>
              <td
                class="stat-name"
              >
                Trouble:
              </td>
              <td
                class="primary-stat-description"
              />
            </tr>
            <tr>
              <td
                class="stat-name"
              >
                Relationship:
              </td>
              <td
                class="primary-stat-description"
              />
            </tr>
            <tr>
              <td
                class="stat-name"
              >
                Aspect One:
              </td>
              <td
                class="primary-stat-description"
              />
            </tr>
            <tr>
              <td
                class="stat-name"
              >
                Aspect Two:
              </td>
              <td
                class="primary-stat-description"
              />
            </tr>
          </tbody>
        </table>
      </article>
      <article
        class="character-sheet-item-secondary"
      >
        <table
          class="padded-table"
        >
          <caption
            class="table-header"
          >
            Approaches
          </caption>
          <tbody>
            <tr>
              <td
                class="stat-name"
              >
                Expected Approach Name:
              </td>
              <td
                class="secondary-stat-description"
              >
                expected appraoch value
              </td>
            </tr>
          </tbody>
        </table>
      </article>
    </section>
    <section
      class="character-stats"
    >
      <article
        class="character-sheet-item-primary"
      >
        <h5
          class="character-sheet-header stunt-header character-sheet-text table-header"
        >
          Stunts
        </h5>
        <ul
          class="stunt-list"
        />
      </article>
    </section>
    <section
      class="character-stats"
    >
      <article
        class="character-sheet-item-primary"
      >
        <table
          class="padded-table"
        >
          <caption
            class="table-header"
          >
            Consequences
          </caption>
          <tbody>
            <tr>
              <td
                class="stat-name"
              >
                Expected Con Name:
              </td>
              <td
                class="primary-stat-description"
              >
                expected con value
              </td>
            </tr>
          </tbody>
        </table>
      </article>
      <article
        class="character-sheet-item-secondary"
      >
        <table>
          <caption>
            Stress
          </caption>
          <tbody>
            <tr>
              <td>
                <button
                  class="stress-box "
                  type="button"
                >
                  1
                </button>
              </td>
              <td>
                <button
                  class="stress-box "
                  type="button"
                >
                  2
                </button>
              </td>
              <td>
                <button
                  class="stress-box "
                  type="button"
                >
                  3
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </article>
    </section>
  </main>
</DocumentFragment>
`;
